### ISSUE TITLE
bug(crossasset): TrainGPU OOM crashes DGX Spark — unbounded tensor allocation in backward pass

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#314]
## Summary

`TrainGPU` in v1.39.0 crashes the DGX Spark (Grace Hopper, 120GB unified memory) within 15-20 minutes of training, requiring a hard reboot. Tested twice with batch_size=256 and batch_size=32 — both crash. The system becomes unresponsive (SSH timeout, ping works) before the OOM killer terminates the process.

## Blocking

This is the only blocker preventing Wolf (our trading system) from running end-to-end in Go. We currently use Python/PyTorch for training (75% walk-forward accuracy) and need Zerfoo to match this. The CPU-only `Train()` achieves 55% because it only backprops through the head layer. `TrainGPU` with full backprop is the fix — but it crashes.

## Reproduction

```go
m := crossasset.NewModel(crossasset.Config{
    NSources:          12,
    FeaturesPerSource: 37,
    DModel:            128,
    NHeads:            4,
    NLayers:           2,
    LearningRate:      0.001,
})

engine, _ := compute.NewGPUEngine[float32](numeric.Float32Ops{})

// 26,555 samples, 50 epochs, batch_size=32
result, err := m.TrainGPU(data, labels, crossasset.TrainConfig{
    Epochs:       50,
    BatchSize:    32,
    LearningRate: 0.001,
}, engine)
// System hangs after ~15 minutes, never returns
```

Data: 797K 1-minute COIN stock bars aggregated to 12 Fibonacci scales (1-233m), 37 TA features per scale, 26,555 training samples after stride=30 subsampling.

## Suspected root cause

`layerCache` in `gpu_train.go` allocates tensors for every intermediate value at every layer:

```go
type layerCache struct {
    xIn, q, k, v    *tensor.TensorNumeric[float32]
    scores, attn     *tensor.TensorNumeric[float32]
    attnOut, concat  *tensor.TensorNumeric[float32]
    projOut, res1    *tensor.TensorNumeric[float32]
    normed, ffnH     *tensor.TensorNumeric[float32]
    ffnAct, ffnOut   *tensor.TensorNumeric[float32]
    res2             *tensor.TensorNumeric[float32]
}
```

With 2 layers, each forward pass creates ~30 tensors. Over 26K samples / 32 batch = 831 batches × 50 epochs = 41,550 forward+backward passes. If tensors aren't freed between batches, that's millions of GPU allocations.

**Suggested fix:** Pool and reuse tensors between batches. Pre-allocate the `layerCache` once per fold and zero/overwrite on each batch, similar to `patchtst_gpu_train.go` which uses `pre-allocate batch workspace outside training loop`.

## Environment

- DGX Spark (Grace Hopper), linux/arm64
- 120GB unified CPU/GPU memory
- Go 1.26
- Zerfoo v1.39.0, ztensor v1.2.0
- `nvidia-smi` reports `[N/A]` for memory.used (unified memory)

## Impact

Without this fix, Wolf cannot train in Go and must maintain a Python/PyTorch training pipeline alongside the Go inference pipeline. This doubles the codebase complexity and prevents deploying a single Go binary for the trading system.

## Related

- Issue #312 (original request for GPU training — feature was added, this is the stability bug)
- PatchTST GPU training works correctly with similar data sizes — the tensor pooling pattern there is the model to follow
